docs/Security_Enhancement_For_AI_Models.md

```diff
@@ -0,0 +1,249 @@
+- Get the user key from key management server (KMS), generate a symmetric wrapping
+    key (SWK) to encrypt the user key (wrapped_key).
+- Encrypt the SWK by the public key from client (wrapped_swk).
+
+For a key broker client, here is an example flow to get a key from KBS:
+
+- Generate 2048 bit RSA key pair (a public key and a private key).
+- Encode the public key to base64 for transferring (user_data).
+- Get quote in the TEE with the hash of the public key for measurement (quote).
+- Request wrapped_key and wrapped_swk from KBS with quote and user_data.
+- Decrypt the user key by the SWK.
+
+An example interface to communicate to the KBS:
+
+- Request
+
+```Python
+# HTTP Headers
+headers = {
+    "Accept": "application/json",
+    "Content-Type": "application/json",
+    "Attestation-Type": "TDX"
+}
+# HTTP Body
+req_body = {
+    "quote": quote,
+    "user_data": user_data
+}
+```
+
+- Response
+
+```JSON
+{
+    "wrapped_key":"",
+    "wrapped_swk":""
+}
+```
+
+### 2.5 Attestation
+
+CCNP provides an easy way to get the quote for attestation. Before getting the quote, a RSA key pair need to be generated for wrapping purpose and the public key will be measured as the user data input to the quote fetching process.
+
+The public key is used as parameter to get the key, the KBS is to encyrpt the key wrapped key by the public key.
+
+The sample code to get the quote with user data:
+
+```Python
+import base64
+
+from ccnp import Quote
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+private_key = rsa.generate_private_key(public_exponent=65537, key_size=3072)
+pubkey = private_key.public_key()
+pubkey_der = pubkey.public_bytes(encoding=serialization.Encoding.DER,
+                                 format=serialization.PublicFormat.SubjectPublicKeyInfo)
+user_data = base64.b64encode(pubkey_der).decode('utf-8')
+quote = Quote.get_quote(user_data=user_data)
+```
+
+### 2.6 Model Decryption
+
+Just like the encryption, the AES 256 GCM encrypted data format should be:
+
+12 bytes header | [12] bytes IV | encrypted data | [16] bytes tag
+---|---|---|---
+
+and the 12 bytes header:
+
+uint32 IV length | uint32 tag length | uint32 data length
+---|---|---
+
+To decrypt the data, here are some sample codes:
+
+```
+
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+
+def decrypt_data(encrypted_data, key) -> bytes:
+
+    if encrypted_data is None or key is None:
+        raise ValueError("Encrypted data or key is empty")
+
+    header_len = 12
+    iv_len, tag_len, data_len = struct.unpack('<3I', encrypted_data[:header_len])
+    iv = encrypted_data[header_len : (iv_len + header_len)]
+    data = encrypted_data[(iv_len + header_len) : -tag_len]
+    tag = encrypted_data[-tag_len:]
+
+    decryptor = Cipher(algorithms.AES(key), modes.GCM(iv, tag)).decryptor()
+    decrypted_data = decryptor.update(data) + decryptor.finalize()
+    return decrypted_data
+```
+
+# References
```

Comment:
Reference normally points to some documents, is there any general description document about the KBS and its functionalities? So that we could list the link here instead of direct python code:)

Reply:
This document tries to explain the design and implementation of CNAP AI model security, so currently the reference is the source code. And it will be updated after some other project document publication.